### PR TITLE
Reduce memory complexity of LocalOperator by d^K

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR lowers memory complexity of LocalOperator by d^k where d is local Hilbert space dimension and $K$ is the largest K-body term.

Effectively addresses #1555 and several past issues, I think.

To do this I need to find the maximum number of connected elements off diagonal per operator, so I need to scan all terms, and it's a bit complicated to do for sparse matrices. I asked chatgpt to write it so I'm not sure it's the best way, but works and by using sparsity in other places we should get slight speedups...

In Zakari sample I get the following improvement:
```python
In [2]: %timeit op._setup(force=True)
4.17 s ± 33.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [9]: op._x_prime.shape
Out[9]: (666, 1369, 2, 2)

In [10]: op._mels.shape
Out[10]: (666, 1369, 2)
```
vs before
```python
In [2]: %timeit op._setup(force=True)
15.03 s ± 856.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: op._x_prime.shape
Out[3]: (666, 1369, 1368, 2)

In [4]: op._mels.shape
Out[4]: (666, 1369, 1368)
```

Tests pass, but I would like this to be tested better before being merged.

cc @z-denis

### CODE REVIEW

**Comment by Z-Denis** [comment]
This looks great. For bosons it makes a lot of sense. We should maybe test that there is no performance regression when dealing with low local hilbert-space dimensions and many terms, i.e. building some Ising model with LocalOperators, although this use case should probably benefit from the sparse-matrix handling.

**Comment by PhilipVinc** [comment]
Runtime-wise the code is unchanged, slowdowns are purely at compile (`_setup()`) time.

You are right that in small Hilbert dimension case (eg spins) this might be suboptimal but I have not checked yet.

Note that we were already iterating through all rows of all operators, but we were converting the terms to dense matrices before, while now I'm leveraging sparsity. Probably on a 2x2 matrix it's faster to work on the dense matrix...

Maybe there's a size at which it makes sense to switch.

I'm off to holiday for a few weeks, so I won't work on this. If you or anybody manages to get a few benchmarks in in a few extra realistic cases it would be helpful to know if we can merge as is or if we need to add something else

**Comment by Z-Denis** [comment]
> Runtime-wise the code is unchanged, slowdowns are purely at compile (`_setup()`) time.

True, but then who cares about it being suboptimal for spins, there for large systems the simulation is already expected to be long. To be sure, I'll try anyway.

> Note that we were already iterating through all rows of all operators, but we were converting the terms to dense matrices before, while now I'm leveraging sparsity. Probably on a 2x2 matrix it's faster to work on the dense matrix...

It's probably faster to use dense matrices indeed.

Have nice vacations, I can check this in the meantime.

**Comment by gcarleo** [comment]
In my tests on spin models I didn't measure any meaningful slowdown because of this PR, whereas of course the improvement is substantial for bosons. @Z-Denis if you also don't measure slow downs I would propose we just merge it !

**Comment by gcarleo** [comment]
(there are however some tests failing, I suspect you need to rebase )

**Comment by Z-Denis** [comment]
Well, if you checked for spins and observed no slow down then, yeah, let's just merge.

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -36,7 +36,7 @@ jobs:
             python-version: "3.8"
             mpi: ""
             main_tests: true
-            oldest_deps: true
+            oldest_deps: false
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply .github/workflows/CI.yml:39]: The `oldest_deps` test are hard to keep reliably working, and I will remove them. I just disabled them here to limit the noise in the failing tests.

**Comment by PhilipVinc** [comment]
rebased.

**Comment by gcarleo** [comment]
I am OK to merge this if you are (the checks on 3.11 somehow take too long and fail, but it's not this PR, we should address that separately)

**Review by gcarleo — APPROVED** [review]

**Comment by PhilipVinc** [comment]
yeah, that tester has been broken for a little while. No idea why, but they pass locally.

**Pull request merged**